src/init.cpp

```diff
@@ -1841,6 +1841,50 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
         }
     }
 
+#ifdef __APPLE__
+    // Warn about exFAT filesystem usage on macOS
+    struct PathCheck {
+        fs::path path;
+        std::string_view description;
+    };
+
+    std::array<PathCheck, 2> paths{{
+        {args.GetDataDirNet(), "data directory"},
+        {args.GetBlocksDirPath(), "blocks directory"}
+    }};
+
+    std::vector<std::string> exfat_paths;
+    std::vector<std::string> error_paths;
+
+    for (const auto& check : paths) {
+        FSType fs_type = GetFilesystemType(check.path);
+        switch(fs_type) {
+            case FSType::EXFAT:
+                exfat_paths.push_back(strprintf("%s (\"%s\")",
+                    check.description,
+                    fs::PathToString(check.path)));
+                break;
+            case FSType::ERROR:
+                error_paths.push_back(strprintf("%s (\"%s\")",
+                    check.description,
+                    fs::PathToString(check.path)));
+                break;
+            default:
+                break;
+        }
+    }
+
+    if (!exfat_paths.empty()) {
+        InitWarning(strprintf(_("The following paths are on exFAT which is known to have intermittent corruption problems on MacOS: %s. "
```

Comment:
Not just in this PR  I’ve noticed `InitWarning` emits duplicate warnings when running bitcoind:
one with a timestamp and another without.

When running Qt from the command line: one popup warning and one log entry without a timestamp.

Reply:
e.g
```
2025-06-25T06:13:16Z [warning] The following paths are on exFAT which is known to have intermittent corruption problems on MacOS: blocks directory ("/Volumes/StoreJet/blocks_temp/blocks"). Move these directories to a non-exFAT formatted drive to avoid corruption.
Warning: The following paths are on exFAT which is known to have intermittent corruption problems on MacOS: blocks directory ("/Volumes/StoreJet/blocks_temp/blocks"). Move these directories to a non-exFAT formatted drive to avoid corruption.
```

Reply:
I'm pretty sure this is actually just two different warning styles both being logged to the same place by the user. I haven't tested, but this code: https://github.com/bitcoin/bitcoin/blob/ad654a4807cd584be9ffcd8640f628ab40cb5170/src/noui.cpp#L33-L46 logs one error to log (with timestamp) and when running the daemon the same error (without timestamp) to `stderr`.

When running daemon in the foreground, both messages appear printed in the console. But IMO this makes sense for the daemonised process to outpue a message like this not only to the logfile, but also stderr (to be read/processed by other tools).

Reply:
Yes